server/methods/browseChannels.js

```diff
@@ -41,23 +42,158 @@ const sortUsers = function(field, direction) {
 	}
 };
 
-Meteor.methods({
-	browseChannels({ text = '', workspace = '', type = 'channels', sortBy = 'name', sortDirection = 'asc', page, offset, limit = 10 }) {
-		const regex = new RegExp(s.trim(escapeRegExp(text)), 'i');
+const getChannels = (user, canViewAnon, searchTerm, sort, pagination) => {
+	if ((!user && !canViewAnon) || (user && !hasPermission(user._id, 'view-c-room'))) {
+		return;
+	}
 
-		if (!['channels', 'users'].includes(type)) {
-			return;
-		}
+	const result = Rooms.findByNameOrFNameAndType(searchTerm, 'c', {
+		...pagination,
+		sort: {
+			featured: -1,
+			...sort,
+		},
+		fields: {
+			t: 1,
+			description: 1,
+			topic: 1,
+			name: 1,
+			fname: 1,
+			lastMessage: 1,
+			ts: 1,
+			archived: 1,
+			default: 1,
+			featured: 1,
+			usersCount: 1,
+			prid: 1,
+		},
+	});
+
+	return {
+		total: result.count(), // count ignores the `skip` and `limit` options
+		results: result.fetch(),
+	};
+};
 
-		if (!['asc', 'desc'].includes(sortDirection)) {
-			return;
+const getTeams = (user, searchTerm, sort, pagination) => {
+	if (!user) {
+		return;
+	}
+
+	const userSubs = Subscriptions.cachedFindByUserId(user._id).fetch();
+	const ids = userSubs.map((sub) => sub.rid);
+	const result = Rooms.findByNameOrFNameInIdsWithTeams(searchTerm, ids, {
+		...pagination,
+		sort: {
+			featured: -1,
+			...sort,
+		},
+		fields: {
+			t: 1,
+			description: 1,
+			topic: 1,
+			name: 1,
+			fname: 1,
+			lastMessage: 1,
+			ts: 1,
+			archived: 1,
+			default: 1,
+			featured: 1,
+			usersCount: 1,
+			prid: 1,
+			teamId: 1,
+			teamMain: 1,
+		},
+	});
+
+	const rooms = result.fetch();
+	const teamIds = [...new Set(rooms.map((r) => r.teamId))];
+	const teams = Promise.await(Team.listByIds(teamIds, { projection: { _id: 1, name: 1 } }));
```

Comment:
I think we can avoid this since the team name will be same as the channel name.

I wonder if we could actually just perform a find on teams collection though, wdyt?

Reply:
Yeah, I was doing this because the `teams.create` endpoint allows you to pass a custom name to the room, but just noticed it gets always override by the team name. Since we can "guarantee" that the room => team.name, we should even be able to avoid the find at all.

I didn't want to use the teams collection there, so it keeps contained in its service 🤔

Reply:
Done!